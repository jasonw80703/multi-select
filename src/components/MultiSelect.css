.container {
  display: inline-block;
  margin: 0.5em;
}

.dropdown {
  width: 30em;
  position: relative;
  cursor: pointer;
  display: inline-block;
  border-radius: 0.5em;
  padding: 0.5em 1em;
  border: 1px solid lightgray;
}

.no-title {
  height: 1.19em;
}

.dropdown-icon {
  position: absolute;
  right: 1em;
}

.flipped-icon {
  transform: scaleY(-1);
}

.options-list {
  padding: 0.2em;
  margin: 0;
  border: 1px solid lightgray;
  border-top: none;
  border-radius: 0 0 0.5em 0.5em;
  line-height: 2em;
}

.options-list li {
  list-style: none;
  cursor: pointer;
}

li.single-action:hover {
  background-color: lightgray;
}

.batch-action {
  color: white;
  background-color: gray;
  border-radius: 0.5em;
}

.selected-datum {
  background-color: lightgray;
}

.option-checkbox {
  position: absolute;
  display: block;
  margin-top: 0.7em;
}

.label-container {
  padding: 0.2em;
}

.label-caption {
  display: block;
  text-align: left;
  font-size: 0.2em;
}

.no-values-caption {
  color: red;
}