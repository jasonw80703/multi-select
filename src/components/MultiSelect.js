import React, { useState, useMemo } from 'react';
import PropTypes from 'prop-types';
import './MultiSelect.css';
import convertDataToString from '../helpers/convertDataToString';

const MultiSelect = ({
  label,
  multiple,
  title,
  values,
}) => {
  const [showList, setShowList] = useState(false);
  const defaultCheckedData = useMemo(() => values.reduce((agg, val) => {
    return {...agg, [val]: false };
  }, {}), [values]);
  const [checkedDataMap, setCheckedDataMap] = useState(defaultCheckedData);

  const anchorDisplayValue = convertDataToString(checkedDataMap, title);
  const dropdownIconClass = `dropdown-icon ${showList && 'flipped-icon'}`;
  const anySelected = Object.values(checkedDataMap).some(datum => datum === true);

  const selectCheckbox = value => {
    const restOfData = multiple ? checkedDataMap : defaultCheckedData;
    setCheckedDataMap({
      ...restOfData,
      [value]: !checkedDataMap[value],
    });
  }

  // Set all values to negation of `anySelected` state value, so that:
  // if any values were selected, then now all should be deselected
  // if no values were selected, then now all should be selected
  const selectOrDeselectAll = () => {
    setCheckedDataMap(values.reduce((agg, val) => {
      return {...agg, [val]: !anySelected };
    }, {}));
  }

  if (values.length === 0) {
    return (
      <div className='container'>
        <span className='dropdown'>
          {anchorDisplayValue}<span className={dropdownIconClass}>▼</span>
        </span>
        {/* <div className='label-container'>
          <span className='label-caption no-values-caption'>No values provided</span>
        </div> */}
      </div>
    )
  }

  return (
    <div className='container'>
      <span
        className={`dropdown ${!anchorDisplayValue && 'no-title'}`}
        onClick={() => setShowList(!showList)}
        data-testid='dropdown-anchor'
      >
        {anchorDisplayValue}<span className={dropdownIconClass}>▼</span>
      </span>
      {showList &&
        <ul className='options-list' data-testid='options-list'>
          {
            multiple &&
              <li className='batch-action' onClick={selectOrDeselectAll} data-testid='batch-action'>
                {anySelected ? 'Deselect All' : 'Select All'}
              </li>
          }
          {
            values.map(value => (
              <li
                className={
                  `single-action ${checkedDataMap[value]
                    && !multiple
                    && 'selected-datum'}`
                }
                key={value.toString()}
                onClick={() => selectCheckbox(value)}
              >
                {
                  multiple &&
                    <input type='checkbox' className='option-checkbox' checked={checkedDataMap[value]} onChange={() => selectCheckbox(value)} />
                }
                <span>{value.toString()}</span>
              </li>
            ))
          }
        </ul>
      }
      {label &&
        <div className='label-container'>
          <span className='label-caption'>{label}</span>
        </div>
      }
    </div>
  )
};

MultiSelect.propTypes = {
  label: PropTypes.string,
  multiple: PropTypes.bool,
  title: PropTypes.string,
  values: PropTypes.array.isRequired,
};

MultiSelect.defaultProps = {
  multiple: false,
};

export default MultiSelect;
