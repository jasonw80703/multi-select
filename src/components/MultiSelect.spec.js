import { render, screen, fireEvent } from '@testing-library/react';
import MultiSelect from './MultiSelect';

describe('<MultiSelect />', () => {
  const defaultProps = {
    values: ['a', 'b', 'c'],
  };

  it('renders initially as expected', () => {
    const { asFragment } = render(<MultiSelect {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('when the select anchor is clicked', () => {
    it('shows the options list', () => {
      render(<MultiSelect {...defaultProps} />);
      const anchor = screen.getByTestId('dropdown-anchor');
      fireEvent.click(anchor);
      expect(screen.getByTestId('options-list')).toBeInTheDocument();
    });

    describe('when closing the options list', () => {
      it('removes the options list', () => {
        render(<MultiSelect {...defaultProps} />);
        const anchor = screen.getByTestId('dropdown-anchor');
        fireEvent.click(anchor);
        fireEvent.click(anchor);
        expect(screen.queryByTestId('options-list')).not.toBeInTheDocument();
      });
    });
  });

  describe('when title is provided', () => {
    const title = 'letters';

    it('shows title', () => {
      render(<MultiSelect {...defaultProps} title={title} />);
      expect(screen.getByText(title)).toBeInTheDocument();
    });
  });

  describe('when label is provided', () => {
    const label = 'This shows a list of letters';

    it('shows label', () => {
      render(<MultiSelect {...defaultProps} label={label} />);
      expect(screen.getByText(label)).toBeInTheDocument();
    });
  });

  describe('when multiple is true', () => {
    const multiDefaultProps = {
      ...defaultProps,
      multiple: true,
    };

    it('renders as expected', () => {
      const { asFragment } = render(<MultiSelect {...multiDefaultProps} />);
      expect(asFragment()).toMatchSnapshot();
    });

    describe('clicking on select all', () => {
      it('selects all options and displays them joined in the dropdown anchor', () => {
        render(<MultiSelect {...multiDefaultProps} />);
        const anchor = screen.getByTestId('dropdown-anchor');
        fireEvent.click(anchor);
        const selectAll = screen.getByTestId('batch-action');
        fireEvent.click(selectAll);
        expect(anchor.textContent).toEqual(
          `${multiDefaultProps.values.join(', ')}▼`
        );
      });
    });
  
    describe('clicking on deselect all', () => {
      const title = 'letters';

      it('deselects all options', () => {
        render(<MultiSelect {...multiDefaultProps} title={title} />);
        const anchor = screen.getByTestId('dropdown-anchor');
        fireEvent.click(anchor);
        const selectAll = screen.getByTestId('batch-action');
        fireEvent.click(selectAll);
        fireEvent.click(selectAll);
        expect(anchor.textContent).toEqual(`${title}▼`);
      });
    });
  });

  describe('when values is an empty array', () => {
    it('no options list appears', () => {
      render(<MultiSelect values={[]} />);
      expect(screen.queryByTestId('dropdown-anchor')).not.toBeInTheDocument();
    });
  });
});
