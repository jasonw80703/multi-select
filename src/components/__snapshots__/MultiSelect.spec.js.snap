// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSelect /> renders initially as expected 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <span
      class="dropdown no-title"
      data-testid="dropdown-anchor"
    >
      <span
        class="dropdown-icon false"
      >
        ▼
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`<MultiSelect /> when multiple is true renders as expected 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <span
      class="dropdown no-title"
      data-testid="dropdown-anchor"
    >
      <span
        class="dropdown-icon false"
      >
        ▼
      </span>
    </span>
  </div>
</DocumentFragment>
`;
